realtek: backport ECC driver

plappermaul opened the pull request:
Upstream will get support for the Realtek ECC engine with 6.18.
To make use of this in Openwrt

- backport upstream patches
- change config so that ECC will be built for nand subtargets
- define ECC engine in RTL93xx DTS.

Djfe commented on target/linux/realtek/files-6.12/drivers/mtd/nand/ecc-realtek.c:
```diff
@@ -0,0 +1,346 @@
+#define   RTL_ECC_CORR_COUNT	GENMASK(19, 12)
+#define   RTL_ECC_RESULT	BIT(8)
+#define   RTL_ECC_ALL_ONE	BIT(4)
+#define   RTL_ECC_OP_STATUS	BIT(0)
```
what is up with the padding after #define ^^?
shouldn't it always be one space not several
I know nitpicky, but it's easier to read, when all `RTL_` lign up, right?

cpackham-atlnz replied:
They're fields of the register above.

Djfe replied:
ah I get it now, thanks

Djfe commented on target/linux/realtek/files-6.12/drivers/mtd/nand/ecc-realtek.c:
```diff
@@ -0,0 +1,346 @@
+#define   RTL_ECC_READ_BE	BIT(22)
+#define   RTL_ECC_WRITE_BE	BIT(21)
+#define   RTL_ECC_IRQ_ENABLE	BIT(20)
+#define RTL_ECC_DMA_TIGGER 	0x08
```
is this supposed to be called DMA_TRIGGER?

plappermaul commented on target/linux/realtek/files-6.12/Documentation/devicetree/bindings/mtd/realtek,nand-ecc-engine.yaml:
```diff
@@ -0,0 +1,32 @@
+
+properties:
+  compatible:
+    const: realtek,rtl9301-ecc
```
upstream x is a no-no.

plappermaul committed:
realtek: backport ECC driver

Upstream will get support for the Realtek ECC engine with 6.18.
To make use of this in Openwrt

- backport upstream patches
- change config so that ECC will be built for nand subtargets
- define ECC engine in RTL93xx DTS.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>
Link: https://github.com/openwrt/openwrt/pull/19746
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented on target/linux/generic/backport-6.12/402-v6.18-mtd-nand-move-chunk-to-core.patch:
Compilation suceeded on x86/64 too.

```
[markus@openwrt linux-6.12.49]$ pwd
/home/markus/openwrt_test/openwrt/build_dir/target-x86_64_musl/linux-x86_64/linux-6.12.49
[markus@openwrt linux-6.12.49]$ ls -lt
total 298672
drwxr-xr-x.  46 markus markus      4096 Sep 29 16:57 tools
drwxr-xr-x.  24 markus markus      4096 Sep 29 16:50 arch
drwxr-xr-x.  27 markus markus      4096 Sep 29 16:50 sound
drwxr-xr-x.  21 markus markus     20480 Sep 29 16:50 lib
drwxr-xr-x.   4 markus markus     20480 Sep 29 16:50 crypto
-rwxr-xr-x.   1 markus markus 106061464 Sep 29 16:50 vmlinux
-rw-r--r--.   1 markus markus   4305153 Sep 29 16:50 System.map

```

robimarko commented:
Thanks! Rebased on top of main and merged!
